Create one hot embedding

takojunior opened the pull request:
The PR has the following updates to Seq2Pat:
- Add one-hot encoding, by mining patterns locally (csp_local.py) or globally (csp_global.py) in each sequence.
- Add Dichotomic Pattern Mining Notebook to show a pipeline of running DPM on a sample dataset.
- Clean up C++ vector variables `datab_MDD`, `mdd_q` from calling `Build_MDD()`
- Clean up vector variable `attr` in `node_mdd.hpp`
- Clean up vector variables `indic_vec`, `result` from calling `Freq_miner()`
- Add an introduction of DPM notebook to README and the Examples section in Docs

skadio commented on README.md:
```diff
@@ -56,7 +56,13 @@ The library offers various constraint types, including a number of non-monotone
-in the [Jupyter Notebook](https://github.com/fidelity/seq2pat/blob/master/notebooks/usage_example.ipynb).
+in the [Usage Example Notebook](https://github.com/fidelity/seq2pat/blob/master/notebooks/usage_example.ipynb).
+
+Supported by Seq2Pat, we proposed **Dichotomic Pattern Mining** ([X. Wang and S. Kadioglu, 2022](https://arxiv.org/abs/2201.09178)) to analyze the correlations between 
```
May be now we can have two sections 
Quick Start - CSPM and another section Quick Start - DPM 

In the DPM example, you can say seq2pat_pos = .. // seq2pat model for positive  and same for negative 
and focus on the call to DPM and call to get_encodings()

skadio commented on sequential/backend/build_mdd.cpp:
```diff
@@ -8,7 +8,7 @@
 
 //Populates the MDD node vector
-void Popl_nodes(int N, int L, int num_att,vector<int>* max_attrs, vector<int>* min_attrs,
+void Popl_nodes(vector<Node*>* datab_m, vector<Pattern*>* dfs_q, int N, int L, int num_att,vector<int>* max_attrs, vector<int>* min_attrs,
```
curiosity question; are you passing these data structures to save on memory? 

If I use the same seq2pat object twice, say I mine once, and then mine again. would this be problem that it's passed by reference here? Can we test plz that repeated calls to seq2pat work ok, if not tested already. Same for remove a constraint and then running again the same seq2pat model

takojunior replied:
Added a test for the repeated calls to seq2pat, it works fine. Also added a test to call seq2pat without constraint, then add constraint, then remove constraint, the results look fine after each call.

skadio replied:
nice!

skadio commented on sequential/backend/freq_miner.cpp:
```diff
@@ -32,7 +32,8 @@ double mm=0;
 	vector<int>* num_minmax, vector<int>* num_avr, vector<int>* num_med, vector<int>* tot_spn, vector<int>* tot_avr, int theta, int L) {
-	result.clear();
+    result.clear();
+    result.shrink_to_fit();
```
is this shrink_to_fit() new? What does it do, can you say a few words

takojunior replied:
Yes, so I am trying to also clear the vector's capacity. `clear()` removes the elements but the capacity, the allocated memory, would still be there. So here I am doing a little more to clean up the memory.

skadio commented on sequential/backend/node_mdd.hpp:
```diff
@@ -29,7 +29,16 @@ class Node {
+					delete attr[i]->at(j);
+		    delete attr[i];
+		}
+		std::vector<vector<Node*>*>().swap(children);
```
can you add a commentary line on these swap operations? Since it is net-new, it would be good to capture what is the idea behind

takojunior replied:
Yes, I can add a comment to this part. This is one of the ways to free the memory taken by the vector object.  The vector is swapped with an empty vector that has no allocated memory, then this can better deallocate the memory. [Here](https://www.techiedelight.com/delete-vector-free-memory-cpp/) I find some comparison between different ways to deallocate the memory of a vector.

skadio replied:
interesting.. thanks for sharing,  Iwas not aware of the shrink_to_fit() method

skadio commented on sequential/backend/pattern.hpp:
```diff
@@ -80,6 +80,15 @@ class Pattern {									    //Nodes of the pattern MDD
 			delete umed[i];
 		}
 
+		vector<vector<Node*>*>().swap(str_pnt);
```
My understanding is it is safe to delete these once a Pattern is destructed. Let me know if I am wrong

takojunior replied:
Yes, basically I am also trying to clean up the memory of the vector, after pattern is destructed.

skadio commented on sequential/backend/seq2pat.cpp:
```diff
@@ -30,49 +30,67 @@ namespace patterns {
+                                 this->theta,
+                                 this->L);
+
+            // Delete MDD nodes
```
nice!

skadio commented on sequential/backend/seq2pat.cpp:
```diff
@@ -30,49 +30,67 @@ namespace patterns {
     std::vector< std::vector<int> > Seq2pat::mine()
     {
         // Queue of MDD nodes
-        std::vector<Pattern*> mdd_q;
```
can you add a comment here on the idea here? Is it to hold onto a single copy of these objects once mine() is called. And then I assume the API changes below is to pass in these single global objects?

takojunior replied:
Yes, this is to have a single hold of the MDD nodes, then we may remove the global variables in `Build_MDD()` to hold the mdd nodes.

skadio commented on sequential/backend/seq_to_pat.pyx:
```diff
@@ -21,11 +21,14 @@ from libcpp.vector cimport vector
 # assigned to its corresponding c_seq2pat element
 cdef class PySeq2pat:
-    cdef Seq2pat c_seq2pat
+    cdef Seq2pat* c_seq2pat;
```
why a pointer now?

takojunior replied:
The idea here can be referring to a cython doc [page](https://cython.readthedocs.io/en/stable/src/userguide/wrapping_CPlusPlus.html), introducing the wrapper class. I am trying to define it as a pointer and explicitly delete this pointer in the `__dealloc__` method to deallocate the memory.

skadio commented on sequential/dpm.py:
```diff
@@ -0,0 +1,141 @@
+    return aggregation_to_patterns
+
+
+def get_one_hot_encodings(sequences: List[list], patterns: List[list], constraints: Union[list, None] = None,
```
the type check for Constraints. Cannot we make it stronger to be sth like List[BaseConstraint]?

takojunior replied:
Yes, this can be made stronger, such as List[_Constraint]. Previously when this is in utils.py, I find out that there would be a circular import issue. But now the issue is gone after modules are put into other files.

skadio commented on sequential/dpm.py:
```diff
@@ -0,0 +1,141 @@
+
+    Returns
+    -------
+    A data frame having one-hot encoding of sequences using the given patterns.
```
plz add the column names of this DF explicitly in the return

skadio commented on sequential/dpm.py:
```diff
@@ -0,0 +1,141 @@
+
+    df = pd.DataFrame()
+    df['sequence'] = sequences
+    df['seq_ind'] = list(range(len(sequences)))
```
minor comment: why need indexes as a new column? Cannot we use DF indexes? OR they are not the same?

takojunior replied:
Right, indexes would suffice. Just tried to use indexes and it worked.

skadio commented on sequential/dpm.py:
```diff
@@ -0,0 +1,141 @@
+
+        # If window size is given, run the local/approximate model
+        if rolling_window_size:
+            df['feat' + str(i)] = df.apply(lambda row: is_satisfiable_in_rolling(row['sequence'], pattern,
```
similar comment related to the DF indexing above, does not row.index suffice?

takojunior replied:
row.index[0] would do it.

skadio commented on sequential/seq2pat.py:
```diff
@@ -214,6 +226,23 @@ def has_lower_bound(self):
         return self.upper_bound is not None
 
+    def check_satisfaction(self, value):
+        res = [True]
```
say a word on why this needs to start from [True] by default

skadio replied:
I understand. What I meant is to add a comment on the code for why so, for the next person looking at this

takojunior replied:
Have added the comment and also have the logic simplified as suggested by @dorukkilitcioglu .

skadio commented on sequential/seq2pat.py:
```diff
@@ -123,6 +125,9 @@ class _Constants:
     # List where index correspond to the attribute ids and values to the minimum value for that attribute
     min_attrs = 'min_attrs'
 
+    # Minimum row count as a threshold of mined patterns
```
hm.. is this a new parametr? If not, how come we didn't have it before? are we adding a new constraint or sth

takojunior replied:
This is still not used though. I added it here when I make a change to`_get_cython_imp()`, to use the seq2pat backend to implement the encoding, but I didn't choose that route then and this should be removed.

skadio commented on sequential/seq2pat.py:
```diff
@@ -282,7 +332,7 @@ def __init__(self, sequences: List[list]):
 
         # Constraint store: attribute_id -> constraint_name -> Constraint
-        self.attr_to_cts: Dict[int, Dict[str, _Constraint]] = dict()
+        self.attr_to_cts: Dict[Attribute, Dict[str, _Constraint]] = dict()
```
nice!

skadio commented on sequential/seq2pat.py:
```diff
@@ -405,10 +456,10 @@ def get_patterns(self, min_frequency: Num) -> List[list]:
 
         # Frequent mining
-        patterns = self._cython_imp.mine()
+        patterns = self.cython_imp.mine()
```
wait.. why did this become a public object now?

takojunior replied:
This is also a minor change when I work on the memory clean-up and the encoding implementation with seq2pat backend. This could be changed back to private.

skadio commented on CHANGELOG.txt:
```diff
@@ -2,6 +2,18 @@
+- Add Dichotomic Pattern Mining notebook, running on sample data
+- Clean up C++ vector variables datab_MDD, mdd_q from calling Build_MDD
+- Clean up vector variable attr in node_mdd.hpp
+- Clean up vector varaibles indic_vec, result from calling Freq_miner
```
typo variables

dorukkilitcioglu reviewed (commented):
I had a high-level look at the PR. Left a bunch of comments, but 2 major things that I'm wondering:
- Cleaning up some of the API artifacts from testing that Serdar already mentioned
- I'm not the best person to review the CSP approaches. However, I was confused as to why you need to do a recursive search to find the pattern features. Looks like we're pattern mining again, but this time using Python instead of the C++ backend?

I'd be happy to be upskilled in my 2nd point, especially if I'm missing something, but that shouldn't be a blocker. Once the first point gets updated and version is bumped to 1.3, I'm happy to approve.

takojunior commented on sequential/backend/seq2pat.cpp:
```diff
@@ -30,49 +30,67 @@ namespace patterns {
+	        std::vector<vector<int>>().swap(result);
+	        // Delete pointers
+            delete datab_MDD;
+            // mdd_q, the queue should be empty after calling Freq_miner()
```
This could be referring to the backend implementation in [freq_miner](https://github.com/fidelity/seq2pat/blob/create_one_hot_embedding/sequential/backend/freq_miner.cpp#L38). I might not be able to explain the detailed steps, but one observation is that all the objects in this queue will be popped out after running the algorithm. Before the object pointer gets popped, the object gets destroyed manually by calling `~Pattern()`.

takojunior commented on sequential/seq2pat.py:
```diff
@@ -417,6 +468,9 @@ def get_patterns(self, min_frequency: Num) -> List[list]:
         patterns_sorted = sort_pattern(patterns)
 
+        # Clean up memory
+        gc.collect()
```
Yes, this is still helpful to reduce memory usage. The clean-up in backend hasn't been able to solve memory deallocation issue completely.

takojunior commented on sequential/seq2pat.py:
```diff
@@ -224,26 +253,52 @@ def __ge__(self, other):
         def __init__(self, attribute: Attribute):
             super().__init__(attribute)
 
+        def check_satisfaction(self, value):
```
Right, I didn't think of that. Removing these also works. Only Gap constraint can keep the function since it needs some variation.

skadio replied:
please make sure all new functions are used, and drop the unused ones. There is quite a bit change in this PR and if we don't do a good clean-up right now, we might not be courageous enough in the future to remove any of these.

takojunior commented on sequential/seq2pat.py:
```diff
@@ -214,6 +226,23 @@ def has_lower_bound(self):
+    def check_satisfaction(self, value):
+        res = [True]
+
+        if self.has_upper_bound():
```
Right! I don't need to append False to res and then goes to the next constraint. We can just return False. Thanks.

takojunior commented on sequential/seq2pat.py:
```diff
@@ -156,6 +156,18 @@ def values(self):
+        validate_attribute_values(values)
+
+        self._values = values
+        self._max = get_max_value(values)
```
No, it is not no longer used currently. It can be removed.

takojunior commented on sequential/utils.py:
```diff
@@ -190,20 +189,46 @@ def write_items(file_name: str, items: List[list]) -> NoReturn:
     open_file.close()
 
 
-def calc_average(result: List[list]) -> list:
```
No, these were not used and these can be cleaned up. I think these were originally added for tests, but they were not used as tests are more formalized.

dorukkilitcioglu replied:
Cool, thanks for the cleanup.

skadio commented on sequential/dpm.py:
```diff
@@ -0,0 +1,146 @@
+    A data frame having one-hot encoding of sequences using the given patterns.
+    
+    Example
+        sequence      feat0 feat1 feat2 ...
```
Minor comment: I am not sure about the "feat" columns.. Why not spell this out completely, feature, and append an underscore so that from the outside someone can split easily if needed (e..g, grab some subset) 

So I would be explicit here and call this "feature_0", "feature_1" and so on

takojunior replied:
Agree. `feature_` makes it clearer.

takojunior commented on tests/test_dpm.py:
```diff
@@ -0,0 +1,655 @@
+        price_attr_ct_neg = 3 <= price_attr_neg.median() <= 4
+        seq2pat_neg.add_constraint(price_attr_ct_neg)
+
+        sequences = sequences_pos + sequences_neg
```
Sorry for the confusion. I am concatenating the two lists of sequences, not the seq2pat objects. But actually these should be used in encoding rather than in the test for DPM patterns, so these can be dropped.

skadio commented on sequential/seq2pat.py:
```diff
@@ -1,19 +1,21 @@
+    validate_attribute_values, validate_sequences
 
-__version__ = "1.2.2"
+__version__ = "1.2.3"
```
by the way, should not this be 1.3.0 --we are adding a complete new DMP functionality

takojunior replied:
Right, 1.3.0 is reasonable.

takojunior commented on sequential/csp_local.py:
```diff
@@ -0,0 +1,188 @@
+    return res
+
+
+def _meet_constraints_in_rolling(sequence: list, pattern: list, seq_attr_ind: int,
```
The idea makes sense, which is another global solution for mining the patterns in a sequence and it is actually experimented previously, but this solution shows to be much slower than csp_local on large number of sequences, so I I didn't go this route finally.

dorukkilitcioglu replied:
Ah, of course you've experimented with it :) It's counter-intuitive to me that the recursive Python solution is faster, but it is what it is. Thanks for confirming!

takojunior commented on requirements.txt:
```diff
@@ -1 +1,4 @@
\ No newline at end of file
+cython>=0.29.13
+ortools
+numpy
```
No, they were not required before.

takojunior commented on notebooks/dichotomic_pattern_mining.ipynb:
```diff
@@ -0,0 +1,1294 @@
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
+      "DPM finished! Runtime: 12.5875 sec\n",
```
Yes, the process might be straightforward, but they provide different options for being easy to use in downstream tasks.

takojunior commented:
@skadio @dorukkilitcioglu, thanks for the comments. I believe I have all them addressed. Some updates will include the following parts:
- Added a DPM example in README, docs, and the bibtex citation
- Added tests for repeatedly calling seq2pat, with or without constraints
- Updated encodings column names, e.g. `feature_0`
- Added comments to backend where vectors get cleaned up
- Added more description to DPM notebook, updated the print out with new column names
- Removed unused functions, parameters,  such as set_attribute(), simplified logic in checking constraints
- Moved the lines for cleaning up the same vector together
- Set the version=1.3.0

Let me know if these updates look good.

skadio commented:
Sounds good!

dorukkilitcioglu commented on sequential/seq2pat.py:
```diff
@@ -1,19 +1,22 @@
+from sequential._version import __version__
 
-__version__ = "1.2.2"
+__version__ = __version__
```
I'm assuming you don't need this line?

takojunior replied:
Yes, the version information is not used in the script though. It could be ok to drop.
